### PR TITLE
[VM] update VM test runner

### LINKED ISSUE #842: Rewrite the test runner
Placeholder issue to rewrite the test runner. It is too verbose right now and the code itself should be cleaned up.

Key points:

- [x] Some tests are not being ran ([transition tests, except to Homestead to DAO](https://github.com/ethereum/tests/tree/develop/BlockchainTests/TransitionTests))
- [x] ~~Not all data available in tests are being used (`postStateHash` in Blockchain tests)~~ (does not matter: if we verify state roots then this data is checked)
- [x] Explicitly verify that if a test is supposed to fail, it does fail
- [x] ~~VM tests are not being ran (`ethereum/tests/VMTests`)~~ [Moved to BlockchainTests](https://github.com/ethereum/tests/pull/680)
- [x] Verify if the tests in [config](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/vm/tests/config.js) which are marked as "failing" are still failing (and why). Also verify that slow tests are being ran on nightly.
- [x] Check all CLI arguments for `tester.js`. They are not documented and most seem to be of (little) use.
- [x] Make tests run on both LegacyTests and on default tests

### PR DESCRIPTION
**Pull request by jochem-brouwer**
Closes #842, #877, #744

The goal of this PR is to harden the test runner with a focus on the integrity of Blockchain tests. It add support for "transition tests" which verifies that the VM switches correctly from one fork to another. Overview of fixes/additions:

- Tests: error was expected, but no error was raised: no "failing test" was raised
- Not all blocks of tests were sometimes ran
- Fix the filters in the test loader so there's not a lot of spam "no test available, skip test..."
- Add flag `--expected-test-amount=n` to verify at least (>=) `n` assertions run in test (otherwise throw 1 failing test)
- Add flag `--verify-test-amount-alltests`: assume all tests are being ran, test afterwards if enough (>=) assertions are run, otherwise throw a failing test
- Document CLI arguments
- Fix cases where blockchain gets locked in case a function which invokes semaphore throws before unlocking the lock, #877 
- Throw test in case there is a reorg in the test (these tests are now temporarily disabled until we fix this, #879 
- Found a single test which runs into a rather complex Block issue, #878 (refactor block)
- Rewrote the PR CI. `MuirGlacier` simply re-runs the Istanbul tests, so this (almost) is a waste of CI time. (Still runs in Nightly)
- Nightly CI now runs all tests including the transition tests.
- Add VM `_updateOpcodes` to reflect the opcodes of the current Common hardfork
- Track which tests+files fail. If there were any failing tests, then this will be reported after the tests (so if your console cuts of the tests you can still see which failed!)

Summary of this PR in a picture:

image

### CODE REVIEW

**Comment by jochem-brouwer** [comment]
Some notes here: 

`/LegacyTests/` has the following `network` params:

```
  Byzantium,
  Constantinople,
  ConstantinopleFix,
  EIP150,
  EIP158,
  Frontier,
  Homestead
```

Normal tests have;

```
  Istanbul,
  Byzantium,
  Constantinople,
  ConstantinopleFix,
  EIP150,
  EIP158,
  Frontier,
  Homestead,
  ByzantiumToConstantinopleFixAt5,
  EIP158ToByzantiumAt5,
  FrontierToHomesteadAt5,
  HomesteadToDaoAt5,
  HomesteadToEIP150At5
```

**Comment by jochem-brouwer** [comment]
I noticed that `ethereum/tests` has blockchain transition tests, but our VM currently does not support this. I added functionality in `runBlock` which checks if we are at a HF block and if so, then update the Common and update the opcodes of VM. I also need to add this functionality in Tx/Block. This will be a rather complex change: if we import Blocks, for instance, then we can supply a `Common` which is currently set to (for instance) Petersburg while we are importing a Byzantium block. We should then thus (temporarily) set Common to Byzantium (to verify difficulty) and then revert it back. I'm not sure how to do this in a clean way. The same goes for importing Frontier txs if we are at a Homestead fork (Homestead disallows certain txs which were allowed in Frontier).

Another problem is that we currently default the chain to `mainnet`, but this implies that if an user sets the VM to a non-Frontier fork it suddenly force the VM to go back to Homestead at the Homestead block.

File: packages/vm/tests/BlockchainTestsRunner.js
Comment (ryanio) [inline packages/vm/tests/BlockchainTestsRunner.js:31]: small style nit: instead of setting `{ common: options.common }` multiple times here and below you can do `const { common } = options` once then simply pass `{ common }` every time like i did with `hardfork`. no big deal though :)

Comment (jochem-brouwer) [reply packages/vm/tests/BlockchainTestsRunner.js:31]: Good point, thanks 😄

**Comment by jochem-brouwer** [comment]
Updated the tests! Don't merge yet, because some tests are now failing! (These are tests which are expected to error, but did not - we had no tests in place!)

**Comment by jochem-brouwer** [comment]
Since some consensus tests were updated in #853, will wait until that's merged and then see if tests are still failing.

File: packages/vm/tests/BlockchainTestsRunner.js
Comment (holgerd77) [inline packages/vm/tests/BlockchainTestsRunner.js:113]: This can be simplified with [Common.setHardforkByBlockNumber()](https://github.com/ethereumjs/ethereumjs-vm/blob/master/packages/common/src/index.ts#L163)

File: packages/vm/tests/tester.js
Comment (holgerd77) [inline packages/vm/tests/tester.js:30]: Oh, that's nice! 😄

File: packages/vm/tests/BlockchainTestsRunner.js
Comment (jochem-brouwer) [reply packages/vm/tests/BlockchainTestsRunner.js:113]: Yeah there is some weird stuff going on over here, because if I re-instantiate a VM after I've updated Common (assuming that is what you mean here) then suddenly a lot more tests start failing. This used to be the way how most blockchain/state tests passed. 🤔

Comment (jochem-brouwer) [reply packages/vm/tests/BlockchainTestsRunner.js:113]: It's worth investigating why this happens though as it should definitely work if we re-instantiate the VM with a correctly-set Common 👍

**Comment by holgerd77** [comment]
Just going through some issues, if it fits you can also integrate #744 here.

**Comment by jochem-brouwer** [comment]
Will post (and edit) the findings here of the failing tests. The changes for these are too big to fit in this PR.

`file: bcBlockRLPAsList test: BLOCK_difficulty_GivenAsList_Frontier` (fork: Frontier)

Problem: the RLP of the block header contains a difficulty which is a **list** of Buffers, and not a Buffer itself. `header.ts` should throw, but does not throw: since we are now checking for tests which should throw but do not, this test fails. 

Related issues: ethereumjs/organization#56, #683

**Comment by jochem-brouwer** [comment]
Temporarily running the nightly tests here.

**Comment by jochem-brouwer** [comment]
Good news! Blockchain tests pass on all forks!

Blockchain tests: (MuirGlacier has no tests -> re-runs Istanbul tests but now on the MuirGlacier fork)
```
Frontier 5530
Homestead 9100
EIP150 5480
EIP158 5512
Byzantium 20276
Constantinople 43476
ConstantinopleFix 43456
Istanbul 46980
Berlin 44
ByzantiumToConstantinopleFixAt5 4
EIP158ToByzantiumAt5 4
HomesteadToDaoAt5 33
HomesteadToEIP150At5 4
```

State tests:

```
Frontier: 1024
Homestead: 1975
EIP150: 1097
EIP158: 1222
Byzantium: 4754
Constantinople: 10530
ConstantinopleFix: 10525
Istanbul: 10759
Berlin: 1414
ByzantiumToConstantinopleFixAt5: 0
EIP158ToByzantiumAt5: 0
HomesteadToDaoAt5: 0
HomesteadToEIP150At5: 0
```

To do:

Check if MuirGlacier actually runs on a VM which has been set to the MuirGlacier Common
Remove the "skip state..." messages from StateTests
Check why some blockchain tests are configured wrong, i.e.:

**Comment by jochem-brouwer** [comment]
Right, I assumed that the tests would actually run the blocks in order, but this is not the case. It seems like we should support reorgs, which we currently do not support. I am not sure if I would think this is a feature of the client or of the blockchain, I first thought it should be a client feature but I'm starting to lean more to blockchain.

Situation is, for instance in `DaoTransactions_HomesteadToDaoAt5`:

Run block 1-4. Throw on block 5. The head of the chain is here block 4 with hash A.

Run block 1-4. All hashes of these blocks are now different than above (so it is a new chain). Throw on block 5. Head of chain is now block 4 of this chain with hash B.

Insert a new block 5 which now has `parentHash` set to hash A. (This is a valid block)

This behavior is not supported by blockchain currently: the reason is that the blockchains' iterator just gets the next block by increasing the block number. It can, if it is has multiple blocks to choose from, thus sometimes (like in the tests) choose the wrong block.

Will disable these tests here, have added a re-org check in the tests which will throw if it encounters a re-org. We should either add support for re-orgs or permanently disable these tests.

**Comment by jochem-brouwer** [comment]
Data per new skip tests, blockchain, assertion amount:

```
--excludeDir=stTimeConsuming
Frontier 4385
Homestead 6997
TangerineWhistle 4255
SpuriousDragon 4305
Byzantium 15379
Constantinople 17189
Petersburg 17174
Istanbul 19817
Berlin 33
```

```
--dir=GeneralStateTests/stTimeConsuming
Frontier 0
Homestead 0
TangerineWhistle 0
SpuriousDragon 0
Byzantium 0
Constantinople 15561
Petersburg 15561
Istanbul 15561
Berlin 0
```

Note: MuirGlacier is the same as Istanbul so not included.

File: packages/vm/tests/config.js
Comment (holgerd77) [inline packages/vm/tests/config.js:285]: TBH I can not follow what this code actually does respectively why it is necessary to rebuild the whole hardfork stack in such a laborious manner when at the end the test is executed on the hardfork set with `const mainnetCommon = new Common('mainnet', hfName)` anyhow?

File: packages/vm/tests/BlockchainTestsRunner.js
Comment (holgerd77) [inline packages/vm/tests/BlockchainTestsRunner.js:125]: Not completely getting this re-org scenario: is this actually triggered somewhere in the tests? Then some of the tests should fail according to the code here I would assume? Or is this a "just in case" implementation? 😛

**Review by holgerd77 — COMMENTED** [review]
Some thing to clarify, otherwise looks good, thanks Jochem, great PR! 😄

File: packages/vm/tests/BlockchainTestsRunner.js
Comment (jochem-brouwer) [reply packages/vm/tests/BlockchainTestsRunner.js:125]: The problem is that if you disable this, most tests pass (see the skipped tests which have a few tests which are added), but not all of these blocks are evaluated (these are the re-org blocks). So the test is only executed partially. There is only one test which actually fails due to this. I just don't like that tests are only partially executed. One of the features of these tests seems to be that it does this reorg stuff, but we are not testing this, so I do not think we should just allow the tests and make them pass while we know that they are only partially executed.

It is rather cumbersome that these tests setup the canonical chain, then do some reorg, and then basically "reorg back" to the initial canonical chain, which means that the tip block of the chain in our implementation is also the tip block which is expected by the test suite, which is also the reason why these tests pass. The blocks in the reorg are not evaluated. If not clear let me know 😄  Related issue which tracks this problem: #879

File: packages/vm/tests/config.js
Comment (jochem-brouwer) [reply packages/vm/tests/config.js:285]: I semi agree that this is a lot of code which does not do a lot. But if we use `mainnetCommon = new Common('mainnet', hfName)`, this means that we are (for instance) running Istanbul on block 0 while Common actually says that we should run Frontier. I do not think that is right, this is to semi future-proof the tests.

Comment (holgerd77) [reply packages/vm/tests/config.js:285]: I think you might be hanging a bit on this block-number thing too much 🙂 , Common just has this information on the underlying block numbers but is not enforcing it in any ways. And Common is used in non-block-number-dependent contexts all the time, e.g. also the VM needs to keep this as one version of using it by e.g. people expecting a Byzantium VM and then feeding it with test blocks where the number plays no role at all (and it should still  behave as Byzantium).

Anyhow, I have the impression we generally need to give this some more reflection, where and how a HF is reset and who is taking responsibility here, especially when moving over to client development.

For now we might want to keep this code, eventually we can remove later depending on the design decision we take along these questions on the VM.

**Review by holgerd77 — APPROVED** [review]
LGTM

**Pull request merged**
